### PR TITLE
ziggy-js support

### PR DESCRIPTION
**Pull request by benallfree**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/ziggy-js/index.d.ts
Comment (danger-public) [inline types/ziggy-js/index.d.ts:3]: - :no_entry_sign: This line should have your github username in it, not /me

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [inline types/ziggy-js/index.d.ts:11]: If I'm understanding this correctly, `Params` here would be `event` in `route('events.show', event)`, and `Param` would be one of the values inside the `event` object?

If that's the case, the types should be slightly different: `event` (`interface Params`) can be a string, a number, an array (of strings, numbers, or objects), or an object. If it's an object, each value of the object (`type Param`) can be a string, number, object, null, or undefined.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [inline types/ziggy-js/index.d.ts:15]:
```suggestion
    methods: Array<'GET' | 'HEAD' | 'POST' | 'PATCH' | 'PUT' | 'OPTIONS' | 'DELETE'>;
```

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [inline types/ziggy-js/index.d.ts:16]: Can also be `undefined` or missing.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [inline types/ziggy-js/index.d.ts:17]: This object can also optionally have a `middleware` key with an array value and a `bindings` key with an object value where all the keys and values of that object are strings.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [inline types/ziggy-js/index.d.ts:24]: This should probably just be `string` because it can be any valid URL protocol, including `ssh`, `ftp`, and tons of others.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [inline types/ziggy-js/index.d.ts:26]: This should be a number, null, or undefined.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [inline types/ziggy-js/index.d.ts:27]: This will be an object, all of its keys and values are strings.

**Comment by bakerkretzmar** [comment]
Ziggy maintainer here—thanks for putting this together @benallfree! Left some comments, hope they make sense given my limited Typescript knowledge! Let me know if you have any questions.

File: types/ziggy-js/index.d.ts
Comment (benallfree) [reply types/ziggy-js/index.d.ts]: I agree with most of this except I don't think an array of objects is valid because `encodeURIComponent` will properly encode an array but not an object.

However, the Typescript definition for `encodeURIComponent` is as follows:

`declare function encodeURIComponent(uriComponent: string | number | boolean): string;`

Since `hydrateUrl` reduces to `encodeURIComponent` for each value, I think the "Typescript" way would be to limit the inputs to compatible types as defined in the official Typescript binding found in `lib.es5.d.ts`.

I've updated it to match.

File: types/ziggy-js/index.d.ts
Comment (benallfree) [reply types/ziggy-js/index.d.ts:29]: Can you clarify please, I thought these routes always pointed back to the Laravel API? When would they ever be anything other than http(s)?

File: types/ziggy-js/index.d.ts
Comment (benallfree) [reply types/ziggy-js/index.d.ts:22]: I added them, but can you please point to where this is referenced in the ziggy js? I searched and didn't find those keys anywhere in the repo.

**Review by benallfree — COMMENTED** [review]
@bakerkretzmar Thank you for taking a look at this, I left a few replies 👍

File: types/ziggy-js/index.d.ts
Comment (benallfree) [inline types/ziggy-js/index.d.ts:34]: @bakerkretzmar Just a side comment, in `route.js` I think it might be nice to do 

```
                if (this.ziggy.defaultParameters.hasOwnProperty(keyName)) {
                    defaultParameter = encodeURIComponent(this.ziggy.defaultParameters[keyName]);
                }
```

That way default params could be any encodable type.

File: types/ziggy-js/index.d.ts
Comment (benallfree) [inline types/ziggy-js/index.d.ts:16]: @bakerkretzmar Taking a slightly different approach here, I noticed that we can be a little more specific about param data that has already been normalized. In all cases, it seems that param values must be something that `encodeURIComponent` can accept. In the Typescript world, that is the `EncodeURIParamType`. Notice that `object` types are not allowed (and in fact aren't supported by encodeURIComponent). But also, `array` types are not allowed even though they are technically supported by the encodeURIComponent implementation (at least in Chrome). According to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) really only `string` should be allowed, but other types seem to work. I trust the official Typescript definition.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [reply types/ziggy-js/index.d.ts]: Ziggy inspects the object to find the parameter value and pulls it out, so `encodeURIComponent()` is never passed an object directly. Something like `route('teams.users.show', [2, { name: 'Taylor', id: 12 }])` is allowed, and it'll eventually result in `encodeURIComponent(12)` for that second parameter. Does that make sense?

See https://github.com/tighten/ziggy/blob/0.9.x/tests/js/test.route.js#L223-L231 and https://github.com/tighten/ziggy/blob/0.9.x/src/js/route.js#L71-L103.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [reply types/ziggy-js/index.d.ts:23]: I don't think `middleware` actually gets used by Ziggy's javascript anywhere, but it can be included with a config option, see https://github.com/tighten/ziggy/blob/0.9.x/src/RoutePayload.php#L98-L104 and https://github.com/tighten/ziggy/blob/0.9.x/tests/Unit/RoutePayloadTest.php#L280-L329.

`bindings` is going to be in `1.0` so you're right actually it's not possible yet, sorry!

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [reply types/ziggy-js/index.d.ts:30]: A user could manually overrides this by modifying the Ziggy config. Now that you mention this though that doesn't really make sense... the only routes in the list would be Laravel named routes. Idk if it's even possible for them to be anything besides `http` and `https`? I'll leave this up to you for now but I'm going to look into this more, if this config option is completely useless we should just remove it.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [reply types/ziggy-js/index.d.ts:35]: Default parameters actually can only be strings and integers, they aren't subject to route model binding: see https://laravel.com/docs/8.x/urls#default-values at the "URL Defaults & Middleware Priority" heading, and https://github.com/laravel/framework/blob/8.x/src/Illuminate/Routing/RouteUrlGenerator.php#L211-L231

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [reply types/ziggy-js/index.d.ts:16]: Ziggy processes the params internally and will only pass integers and strings to `encodeURIComponent()`, so it is actually valid to pass an array of mixed integers, strings, and objects, or an object containing string keys and object values.

I know that's extremely lenient and hard to type-check, but that's how Ziggy works 😂 

All of the following are allowed as `params` to the `route()` function:

```js
1
'foo'
[1]
[1, 2]
['foo', 3]
{ city: 'Toronto', user: 'Jacob' }
{ city: 'Toronto', user: { id: 12, name: 'Jacob' } }
{ id: 1, name: 'Jacob' }
[{ id: 1, name: 'Jacob' }]
[{ id: 1, name: 'Jacob' }, 'foo']
[{ id: 1, name: 'Jacob' }, { id: 2, name: 'Ben' }]
[{ id: 3, name: 'Jacob', details: [/* literally anything is allowed here, it all gets thrown away */] }]
```

File: types/ziggy-js/index.d.ts
Comment (benallfree) [reply types/ziggy-js/index.d.ts]: Hi thanks for those links. I'm sorry if I'm being slow on this, but looking at the `route.js` source it appears that an object will only work as a param value when the object contains a key named `id` https://github.com/tighten/ziggy/blob/0.9.x/src/js/route.js#L99, otherwise it is assumed to be a plain value that can be accepted by `encodeURIComponent`. Do you agree?

I've taken another shot at defining the Params structure accurately, please have another look :)

File: types/ziggy-js/index.d.ts
Comment (benallfree) [reply types/ziggy-js/index.d.ts:27]: Hmm ok I think it is best to omit these for now, at least until 1.0 comes along. Typescript doesn't care if you give it extra keys, so it won't hurt anything if someone adds `{ middleware: 'foo' }` or even `{ bar: 'baz' }`

Does that sound ok to you? People can always include extra keys in TS, it will just treat them as `any` types.

File: types/ziggy-js/index.d.ts
Comment (benallfree) [reply types/ziggy-js/index.d.ts:34]: Ok will leave as is for now. If someone complains I can always update it once we have a concrete use case to support it.

File: types/ziggy-js/index.d.ts
Comment (benallfree) [reply types/ziggy-js/index.d.ts:39]: Ok I've updated it to `string | number`. Is the goal to mimic the PHP `route()` as much as possible? TS doesn't have an `int` type, so the best we can do is `number`. Hope that works :)

File: types/ziggy-js/index.d.ts
Comment (benallfree) [reply types/ziggy-js/index.d.ts:22]: I think we have this dialed in this time 👍   It will require the `id` key if it is a param value and will allow any key name if it is a param object. Plus it will allow strings/numbers/arrays thereof.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [reply types/ziggy-js/index.d.ts]: Yep that's right! The way you have it now looks good to me.

This is going to change very soon in `1.0` to allow any key, not just `id`, sorry 🙃 I'll PR a change here to update and will probably tag you for your help looking at it if that's okay.

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [reply types/ziggy-js/index.d.ts:27]: Perfect 👍

File: types/ziggy-js/index.d.ts
Comment (bakerkretzmar) [reply types/ziggy-js/index.d.ts:39]: Yep that's the goal, and `number` is great thanks!

**Comment by benallfree** [comment]
@bakerkretzmar Can you please make sure your review is marked Approved? I see the status says "needs maintainer action". Or if there is anything else to resolve please let me know :)

**Review by bakerkretzmar — APPROVED** [review]

**Comment by bakerkretzmar** [comment]
Yep sorry about that haha

**Review by amcasey — APPROVED** [review]

**Comment by benallfree** [comment]
Ready to merge

**Pull request merged**
